flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1084,6 +1113,32 @@ class ScalarExprLowering {
     }
   }
 
+  fir::ExtendedValue genArrayLit(
+      const Fortran::evaluate::Constant<Fortran::evaluate::SomeDerived> &con) {
+    llvm::SmallVector<mlir::Value> lbounds;
+    llvm::SmallVector<mlir::Value> extents;
+    auto idxTy = builder.getIndexType();
+    for (auto [lb, extent] : llvm::zip(con.lbounds(), con.shape())) {
+      lbounds.push_back(builder.createIntegerConstant(getLoc(), idxTy, lb - 1));
+      extents.push_back(builder.createIntegerConstant(getLoc(), idxTy, extent));
+    }
+    fir::SequenceType::Shape shape;
+    shape.append(con.shape().begin(), con.shape().end());
+    auto recTy = converter.genType(con.GetType().GetDerivedTypeSpec());
+    auto arrayTy = fir::SequenceType::get(shape, recTy);
+    mlir::Value array = builder.create<fir::UndefOp>(getLoc(), arrayTy);
+    Fortran::evaluate::ConstantSubscripts subscripts = con.lbounds();
+    do {
```

Comment:
Maybe a `for (subs = ...; true; incr(subs))` loop instead of a `do while`?

Reply:
I just took over the pattern from the `genArrayLit`. With, the for option, I am sure there must be a way, but I can't get get it to stop without doing something like:
```
    bool b = true;
    for(auto subscripts = con.lbounds(); b; b = con.IncrementSubscripts(subscripts)) {
```
And I prefer the `do while` to that.